pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java

```diff
@@ -671,8 +671,15 @@ public void updateSubscribeRateLimiter() {
     }
 
     private void asyncAddEntry(ByteBuf headersAndPayload, PublishContext publishContext) {
-        ledger.asyncAddEntry(headersAndPayload,
-            (int) publishContext.getNumberOfMessages(), this, publishContext);
+        // Retain the buffer in advance to avoid the buffer might have been released when it's passed to `asyncAddEntry`
+        final var buffer = headersAndPayload.retain();
```

Comment:
Just wondering if this is going to add an extra reference compared to the previous code. 
I have a solution in #23956 to reliably detect Netty ByteBuf leaks so those changes are one way to validate whether there's a new leak.

Reply:
Oh yes, it's a memory leak. Let me think how to handle it better.

Reply:
Oh, I realized that I made a regression in https://github.com/apache/pulsar/pull/23983/commits/61dcac775d35e4abfada6af749ae7cdf66405248